File: Data_Structures/Linked-List/C/linkedList.c
#include <stdio.h>
#include <stdlib.h>
#include "linkedList.h"



// Função para criar um novo nó
Node* newNode(int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    if (node == NULL) {
        printf("Erro: Não foi possível alocar memória para o novo nó.\n");
        exit(1);
    }
    node->data = data;
    node->next = NULL;
    return node;
}

// Função para criar uma nova lista encadeada
LinkedList* newLinkedList() {
    LinkedList* list = (LinkedList*)malloc(sizeof(LinkedList));
    if (list == NULL) {
        printf("Erro: Não foi possível alocar memória para a nova lista encadeada.\n");
        exit(1);
    }
    list->head = NULL;
    return list;
}

// Função para adicionar um novo nó no final da lista
void appendNode(LinkedList* list, int data) {
    Node* node = newNode(data);
    if (list->head == NULL) {
        list->head = node;
        return;
    }
    struct Node* temp = list->head;

    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = node;
}

// Função para adicionar um novo nó no início da lista
void pushNode(LinkedList* list, int data) {
    Node* node = newNode(data);
    Node* temp = list->head;
    list->head = node;
    node->next = temp;
}

// Função para adicionar um novo nó em uma posição específica da lista
void appendNodeIndex(LinkedList* list, int data, int index) {
    if (index < 0) {
        printf("Erro: O índice não pode ser negativo.\n");
        return;
    }
    if (index == 0) {
        pushNode(list, data);
        return;
    }
    
    Node* nodeData = newNode(data);
    Node* temp = list->head;
    for (int i = 0; i < index -2; i++) {
        if (temp == NULL) {
            printf("Erro: O índice é maior que o tamanho da lista.\n");
            return;
        }
        temp = temp->next;
    }
    if (temp == NULL) {
        printf("Erro: O índice é maior que o tamanho da lista.\n");
        return;
    }
    nodeData->next = temp->next;
    temp->next = nodeData;
}

// Função para retornar o tamanho da lista encadeada
int size(LinkedList* list) {
    Node* temp = list->head;
    int count = 0;
    while (temp != NULL) {
        count++;
        temp = temp->next;
    }
    return count;
}

// Função para imprimir os elementos da lista encadeada
void printList(LinkedList* list) {
    Node* temp = list->head;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

File: Data_Structures/Linked-List/Python/LinkedList.py

from node import Node

class LinkedList:

    # Construtor
    def __init__(self):
        self.head = None

    # Método para adicionar um novo nó no final da lista
    def adicionarFim(self, data):
        newNode = Node(data) 

        if not self.head:
            self.head = newNode
            return

        current = self.head

        while current.next:
            current = current.next
        current.next = newNode

    # Método para adicionar um novo nó no início da lista
    def adicionarInicio(self, data):
        newNode = Node(data) 
        newNode.next = self.head
        self.head = newNode

    # Método para adicionar um novo nó em uma posição específica
    def inserir(self, data, posicao):
        if posicao == 0:
            self.adicionarInicio(data)
            return
        if posicao >= self.tamanho():
            self.adicionarFim(data)
            return

        newNode = Node(data) # Cria um novo nó com os dados fornecidos
        current = self.head
        for i in range(posicao - 1):
            current = current.next
        newNode.next = current.next
        current.next = newNode

    # Método para remover um nó da lista
    def remover(self, data):
        if self.head.data == data:
            self.head = self.head.next
            return

        current = self.head
        while current.next:
            if current.next.data == data:
                current.next = current.next.next
                return
            current = current.next

    # Método para retornar o tamanho da lista
    def tamanho(self):
        count = 0
        current = self.head
        while current:
            count += 1
            current = current.next
        return count    
    
    # Método para imprimir os elementos da lista
    def Exibir(self):
        current = self.head
        while current:
            print(current.data)
            current = current.next
        print()

